File: Leetcode/algorithms/hard/0025_reverse_nodes_in_k_groups/0025_reverse_nodes_in_k_groups.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int val = 0, ListNode next = null) {
        this.val = val;
        this.next = next;
    }
}

public class Solution {
    public ListNode ReverseKGroup(ListNode head, int k) {
        ListNode Reverse(ListNode start, ListNode end) {
            ListNode prev = null;
            ListNode current = start;
            while (current != end) {
                ListNode nextNode = current.next;
                current.next = prev;
                prev = current;
                current = nextNode;
            }
            return prev;
        }

        int count = 0;
        ListNode current = head;
        while (current != null && count < k) {
            current = current.next;
            count++;
        }

        if (count < k) {
            return head;
        }

        ListNode reversedHead = Reverse(head, current);
        head.next = ReverseKGroup(current, k);

        return reversedHead;
    }
}

File: Leetcode_past/algorithms/hard/0025_reverse_nodes_in_k_groups/0025_reverse_nodes_in_k_groups.py
# Definition for singly-linked list.
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def reverse(start, end):
            prev = None
            current = start
            while current != end:
                next_node = current.next
                current.next = prev
                prev = current
                current = next_node
            return prev
        
        count = 0
        current = head
        while current and count < k:
            current = current.next
            count += 1
        
        if count < k:
            return head
        
        reversed_head = reverse(head, current)
        head.next = self.reverseKGroup(current, k)
        
        return reversed_head
